File: opengl2/src/main/java/com/fanfan/opengl/geometry/render/Cube.java
package com.fanfan.opengl.geometry.render;

import android.opengl.GLES20;
import android.opengl.Matrix;
import android.view.View;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.FloatBuffer;
import java.nio.ShortBuffer;

import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;

public class Cube extends Shape {

    private FloatBuffer vertexBuff;
    private FloatBuffer colorBuff;
    private ShortBuffer indexBuff;

    private float cubeCooders[] = {
        -1.0f,1.0f,1.0f, //正面左上角
        -1.0f,-1.0f,1.0f,
        1.0f,-1.0f,1.0f,
        1.0f,1.0f,1.0f,
        -1.0f,1.0f,-1.0f,
        -1.0f,-1.0f,-1.0f,
        1.0f,-1.0f,-1.0f,
        1.0f,1.0f,-1.0f
    };

    private float colors[] = {
      1.0f,0f,0f,1.0f,
            1.0f,0f,0f,1.0f,
            1.0f,0f,0f,1.0f,
            1.0f,0f,0f,1.0f,
            0.0f,1.0f,0f,1.0f,
            0.0f,1.0f,0f,1.0f,
            0.0f,1.0f,0f,1.0f,
            0.0f,1.0f,0f,1.0f
    };

    private short index[] = {
        0,3,2,0,2,1,
        0,1,5,0,5,4,
        0,7,3,0,4,7,
        6,7,4,6,5,4,
        1,5,6,1,2,6,
        3,2,6,3,7,6
    };


    private final String vertexCode =
            "attribute vec4 vPosition;"+
             "uniform mat4 mMatrix;"+
              "varying vec4 vColor;"+
              "attribute vec4 aColor;"+
            " void main(){"+
            "gl_Position = mMatrix*vPosition;"+
                    "vColor = aColor;}";

    private final String fragmentCode =
            "precision mediump float;"+
             "varying vec4 vColor;"+
            "void main(){"+
            " gl_FragColor = vColor;}";

    private int mProgram;

    public Cube(View view) {
        super(view);
        //顶点坐标数据
        ByteBuffer dd = ByteBuffer.allocateDirect(cubeCooders.length *4);
        dd.order(ByteOrder.nativeOrder());
        vertexBuff = dd.asFloatBuffer();
        vertexBuff.put(cubeCooders);
        vertexBuff.position(0);
        // draw element 数据
        ByteBuffer inbuf = ByteBuffer.allocateDirect(index.length*2);
        inbuf.order(ByteOrder.nativeOrder());
        indexBuff = inbuf.asShortBuffer();
        indexBuff.put(index);
        indexBuff.position(0);
        //颜色数据
        ByteBuffer coBuf = ByteBuffer.allocateDirect(colors.length*4);
        coBuf.order(ByteOrder.nativeOrder());
        colorBuff = coBuf.asFloatBuffer();
        colorBuff.put(colors);
        colorBuff.position(0);

        int vertexShader = loadShader(GLES20.GL_VERTEX_SHADER,vertexCode);
        int fragmentShader = loadShader(GLES20.GL_FRAGMENT_SHADER,fragmentCode);
        mProgram = GLES20.glCreateProgram();
        GLES20.glAttachShader(mProgram,vertexShader);
        GLES20.glAttachShader(mProgram,fragmentShader);
        GLES20.glLinkProgram(mProgram);
    }

    private float[] mProjectMatrix = new float[16];
    private float[] mViewMatrix = new float[16];
    private float[] mMVPMatrix = new float[16];

    @Override
    public void onSurfaceCreated(GL10 gl, EGLConfig config) {
        //开启深度测试
        GLES20.glEnable(GLES20.GL_DEPTH_TEST);

    }

    @Override
    public void onSurfaceChanged(GL10 gl, int width, int height) {
        float ratio = (float) width/height;
        Matrix.frustumM(mProjectMatrix,0,-ratio,ratio,-1,1,3,20);
        Matrix.setLookAtM(mViewMatrix,0,-5.0f,5.0f,10.0f,0f,0f,0f,0f,1.0f,0f);
        Matrix.multiplyMM(mMVPMatrix,0,mProjectMatrix,0,mViewMatrix,0);
    }


    private int mVertexHandle;
    private int mColorHandle;
    private int mMatrixHandle;

    @Override
    public void onDrawFrame(GL10 gl) {
        GLES20.glUseProgram(mProgram);

        mVertexHandle = GLES20.glGetAttribLocation(mProgram,"vPosition");
        GLES20.glEnableVertexAttribArray(mVertexHandle);
        GLES20.glVertexAttribPointer(mVertexHandle,3,GLES20.GL_FLOAT,false,0,vertexBuff);

        mColorHandle = GLES20.glGetAttribLocation(mProgram,"aColor");
        GLES20.glEnableVertexAttribArray(mColorHandle);
        GLES20.glVertexAttribPointer(mColorHandle,4,GLES20.GL_FLOAT,false,0,colorBuff);

        mMatrixHandle = GLES20.glGetUniformLocation(mProgram,"mMatrix");
        GLES20.glUniformMatrix4fv(mMatrixHandle,1,false,mMVPMatrix,0);

        GLES20.glDrawElements(GLES20.GL_TRIANGLES,index.length,GLES20.GL_UNSIGNED_SHORT,indexBuff);
        GLES20.glDisableVertexAttribArray(mVertexHandle);
    }
}

File: opengl2/src/main/java/com/fanfan/opengl/vary/Cube.kt
package com.fanfan.opengl.vary

import android.content.res.Resources
import android.opengl.GLES20
import android.util.Log
import com.fanfan.opengl.utils.ShaderUtils
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.FloatBuffer
import java.nio.ShortBuffer

class Cube(var res:Resources) {



    private val cubePositions = arrayOf(
        -1.0f,1.0f,1.0f,
        -1.0f,-1.0f,1.0f,
        1.0f,-1.0f,1.0f,
        1.0f,1.0f,1.0f,
        -1.0f,1.0f,-1.0f,
        -1.0f,-1.0f,-1.0f,
        1.0f,-1.0f,-1.0f,
        1.0f,1.0f,-1.0f
    );

    private val index = shortArrayOf(
        6, 7, 4, 6, 4, 5,    //后面
        6, 3, 7, 6, 2, 3,    //右面
        6, 5, 1, 6, 1, 2,    //下面
        0, 3, 2, 0, 2, 1,    //正面
        0, 1, 5, 0, 5, 4,    //左面
        0, 7, 3, 0, 4, 7    //上面
    )
    private val color  = floatArrayOf(
        0f,1f,0f,1f,
        0f,1f,0f,1f,
        0f,1f,0f,1f,
        0f,1f,0f,1f,
        1f,0f,0f,1f,
        1f,0f,0f,1f,
        1f,0f,0f,1f,
        1f,0f,0f,1f
    )

    private lateinit var vertexBuf:FloatBuffer
    private lateinit var colorBuf:FloatBuffer
    private lateinit var indexBuf:ShortBuffer
      var mProgram:Int = 0
      var hVertex:Int = 0
      var hColor:Int = 0
      var hMatrix:Int = 0

    init {
        initData()
    }

    private fun initData(){
        Log.i("fanfan","cubePositions is ${cubePositions} and res is ${res}")
        val a = ByteBuffer.allocateDirect(cubePositions.size*4)
        a.order(ByteOrder.nativeOrder())
        vertexBuf = a.asFloatBuffer()
        vertexBuf.put(cubePositions.toFloatArray())
        vertexBuf.position(0)
        val b= ByteBuffer.allocateDirect(color.size*4)
        b.order(ByteOrder.nativeOrder())
        colorBuf = b.asFloatBuffer()
        colorBuf.put(color)
        colorBuf.position(0)
        val c = ByteBuffer.allocateDirect(index.size*2)
        c.order(ByteOrder.nativeOrder())
        indexBuf = c.asShortBuffer()
        indexBuf.put(index)
        indexBuf.position(0)
    }

    public fun create(){
        mProgram = ShaderUtils.createProgram(res,"vary/vertex.glsl","vary/fragment.glsl")
        hVertex = GLES20.glGetAttribLocation(mProgram,"vPosition")
        hColor = GLES20.glGetAttribLocation(mProgram,"aColor")
        hMatrix = GLES20.glGetUniformLocation(mProgram,"vMatrix")
    }

    private var matrix:FloatArray? = null
    public fun setMatrix(matrix:FloatArray){
        this.matrix = matrix
    }

    public fun drawSelf(){
        GLES20.glUseProgram(mProgram)
        if (matrix != null) GLES20.glUniformMatrix4fv(hMatrix,1,false,matrix,0)
        GLES20.glEnableVertexAttribArray(hVertex)
        GLES20.glEnableVertexAttribArray(hColor)
        GLES20.glVertexAttribPointer(hVertex,3,GLES20.GL_FLOAT,false,0,vertexBuf)
        GLES20.glVertexAttribPointer(hColor,4,GLES20.GL_FLOAT,false,0,colorBuf)
        //索引法绘制正方形
        GLES20.glDrawElements(GLES20.GL_TRIANGLES,index.size,GLES20.GL_UNSIGNED_SHORT,indexBuf)
        GLES20.glDisableVertexAttribArray(hVertex)
        GLES20.glDisableVertexAttribArray(hColor)
    }

}